# Name: Ajit A Gouda

# Batch:DSG0523

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import  warnings
warnings.filterwarnings('ignore')
import sqlite3
import csv

In [7]:
df=pd.read_csv('42_district.csv')
df

STATE/UT       DISTRICT  Year  Rape  Kidnapping and Abduction  \
0     ANDHRA PRADESH       ADILABAD  2001    50                        30   
1     ANDHRA PRADESH      ANANTAPUR  2001    23                        30   
2     ANDHRA PRADESH       CHITTOOR  2001    27                        34   
3     ANDHRA PRADESH       CUDDAPAH  2001    20                        20   
4     ANDHRA PRADESH  EAST GODAVARI  2001    23                        26   
...              ...            ...   ...   ...                       ...   
9012     LAKSHADWEEP    LAKSHADWEEP  2012     0                         0   
9013     LAKSHADWEEP          TOTAL  2012     0                         0   
9014      PUDUCHERRY       KARAIKAL  2012     6                         2   
9015      PUDUCHERRY     PUDUCHERRY  2012     7                        14   
9016      PUDUCHERRY          TOTAL  2012    13                        16   

      Dowry Deaths  Assault on women with intent to outrage her modesty  \
0               16                                                149     
1                7                                                118     
2               14                                                112     
3               17                                                126     
4               12                                                109     
...            ...                                                ...     
9012             0                                                  1     
9013             0                                                  1     
9014             0                                                  2     
9015             0                                                  7     
9016             0                                                  9     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                             34                                  175   
1                             24                                  154   
2                             83                                  186   
3                             38                                   57   
4                             58                                  247   
...                          ...                                  ...   
9012                           0                                    1   
9013                           0                                    1   
9014                           0                                    1   
9015                           2                                    5   
9016                           2                                    6   

      Importation of Girls  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  
...                    ...  
9012                     0  
9013                     0  
9014                     0  
9015                     0  
9016                     0  

[9017 rows x 10 columns]

## 3.1) Insert records from 42_District_wise_crimes_committed_against_women_2001_2012.csv into a table

In [8]:
db=sqlite3.connect('crime_database.db')
cursor=db.cursor()
cursor.execute("CREATE TABLE Crimes_against_women(State_UT text, District text, Year int, Rape int, Kidnapping_abduction int, Dowry_Deaths int,Assault_on_women_with_intent_to_outrage_her_modesty int,Insult_to_modesty_of_Women int,Cruelty_by_Husband_or_His_Relatives int,Importation_of_Girls int)")

In [9]:
with open(r"42_district.csv") as file:
    records=0
    for row in file:
        cursor.execute('INSERT INTO Crimes_against_women VALUES(?,?,?,?,?,?,?,?,?,?)',row.split(','))
        db.commit()
        records +=1
print(records,"record(s) printed")

9018 record(s) printed


### we have inserted our values into the table Crimes_against_women

## 3.2) Write SQL query to find the highest number of rapes & Kidnappings that happened in which state, District, and year.

In [10]:
result=cursor.execute("SELECT State_UT, District, Year, Rape, Kidnapping_abduction FROM Crimes_against_women WHERE District NOT LIKE '%Total%'ORDER BY (Rape + Kidnapping_abduction) DESC LIMIT 1;")
result.fetchall()

[('WEST BENGAL', 'MURSHIDABAD', 2010, 526, 441)]

#### The data seems to indicate that Murshidabad district in West Bengal has the highest number of reported rape cases (526) and kidnapping cases (441)

## 3.3)Write SQL query to find All the lowest number of rapes & Kidnappings that happened in which state, District, and year

In [11]:
result=cursor.execute("SELECT State_UT, District, Year, Rape, Kidnapping_abduction FROM Crimes_against_women WHERE ( Rape, Kidnapping_abduction) IN ( SELECT MIN(Rape) AS Lowest_Rape,MIN(Kidnapping_abduction) AS Lowest_Kidnapping FROM Crimes_against_women WHERE District NOT LIKE '%Total%')")
result.fetchall()

[('ANDHRA PRADESH', 'GUNTAKAL RLY.', 2001, 0, 0),
 ('ANDHRA PRADESH', 'SECUNDERABAD RLY.', 2001, 0, 0),
 ('ASSAM', 'C.I.D.', 2001, 0, 0),
 ('BIHAR', 'ARWAL', 2001, 0, 0),
 ('CHHATTISGARH', 'GRP RAIPUR', 2001, 0, 0),
 ('HIMACHAL PRADESH', 'G.R.P.', 2001, 0, 0),
 ('JAMMU & KASHMIR', 'CRIME JAMMU', 2001, 0, 0),
 ('JAMMU & KASHMIR', 'CRIME SRINAGAR', 2001, 0, 0),
 ('JAMMU & KASHMIR', 'RAILWAYS', 2001, 0, 0),
 ('MADHYA PRADESH', 'JABALPUR RLY.', 2001, 0, 0),
 ('MAHARASHTRA', 'PUNE RLY.', 2001, 0, 0),
 ('MANIPUR', 'SENAPATI', 2001, 0, 0),
 ('MANIPUR', 'TAMENGLONG', 2001, 0, 0),
 ('MIZORAM', 'MAMIT', 2001, 0, 0),
 ('NAGALAND', 'KIPHIRE', 2001, 0, 0),
 ('NAGALAND', 'PEREN', 2001, 0, 0),
 ('NAGALAND', 'TUENSANG', 2001, 0, 0),
 ('PUNJAB', 'G.R.P', 2001, 0, 0),
 ('SIKKIM', 'NORTH', 2001, 0, 0),
 ('TAMIL NADU', 'CHENNAI RLY.', 2001, 0, 0),
 ('TAMIL NADU', 'TRICHY RLY.', 2001, 0, 0),
 ('WEST BENGAL', 'ASANSOL', 2001, 0, 0),
 ('WEST BENGAL', 'SILIGURI G.R.P.', 2001, 0, 0),
 ('A & N ISLANDS', 'NICOBA

#### The data provided identifies the states and their corresponding districts that had the lowest reported incidents of rape and kidnapping in a given year.

## 3.4 ) Insert records from 02_District_wise_crimes_committed_against_ST_2001_2012.csv into a new table

In [12]:
st=pd.read_csv('St.csv')
st

STATE/UT       DISTRICT  Year  Murder  Rape  Kidnapping Abduction  \
0     ANDHRA PRADESH       ADILABAD  2001       0     1                     2   
1     ANDHRA PRADESH      ANANTAPUR  2001       0     0                     0   
2     ANDHRA PRADESH       CHITTOOR  2001       0     0                     0   
3     ANDHRA PRADESH       CUDDAPAH  2001       0     0                     0   
4     ANDHRA PRADESH  EAST GODAVARI  2001       0     0                     0   
...              ...            ...   ...     ...   ...                   ...   
9013     LAKSHADWEEP    LAKSHADWEEP  2012       0     0                     0   
9014     LAKSHADWEEP          TOTAL  2012       0     0                     0   
9015      PUDUCHERRY       KARAIKAL  2012       0     0                     0   
9016      PUDUCHERRY     PUDUCHERRY  2012       0     0                     0   
9017      PUDUCHERRY          TOTAL  2012       0     0                     0   

      Dacoity  Robbery  Arson  Hurt  Protection of Civil Rights (PCR) Act  \
0           0        0      0     2                                     0   
1           0        0      0     7                                     0   
2           0        0      0     2                                     0   
3           0        0      0     2                                     0   
4           0        0      0     0                                     0   
...       ...      ...    ...   ...                                   ...   
9013        0        0      0     0                                     0   
9014        0        0      0     0                                     0   
9015        0        0      0     0                                     0   
9016        0        0      0     0                                     0   
9017        0        0      0     0                                     0   

      Prevention of atrocities (POA) Act  Other Crimes Against STs  
0                                      0                        13  
1                                      1                         6  
2                                      0                         0  
3                                      2                         0  
4                                      0                        14  
...                                  ...                       ...  
9013                                   0                         0  
9014                                   0                         0  
9015                                   0                         0  
9016                                   0                         0  
9017                                   0                         0  

[9018 rows x 13 columns]

In [13]:
cursor.execute("CREATE TABLE crime_against_st (state_UT TEXT,District TEXT,Year INT,Murder INT,Rape INT,Kidanpping_abduction INT, Dacoity int, Roberry int, Arson int, Hurt int, Protection of Civil Rights Act int, Prevention of atrocities Act int, Other Crimes Against STs int)")

In [14]:
with open(r"st.csv") as file:
    records=0
    for row in file:
        cursor.execute('INSERT INTO crime_against_st VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?)',row.split(','))
        db.commit()
        records +=1

print(records,"record(s) printed")

9019 record(s) printed


## 3.5) Write SQL query to find the highest number of dacoity/robbery in which district.

In [15]:
result=cursor.execute("SELECT state_UT, District,Year, MAX(Dacoity) AS Highest_Dacoity FROM crime_against_st GROUP BY state_UT, District, Year ORDER BY Highest_Dacoity DESC LIMIT 2")
result.fetchall()

[('STATE/UT', 'DISTRICT', 'Year', 'Dacoity'), ('GUJARAT', 'DAHOD', 2004, 29)]

#### The above state and district shows the highest number of dacoity/roberry in that particular year

## 3.6) Write SQL query to find in which districts(All) the lowest number of murders happened

In [16]:
result=cursor.execute("SELECT State_UT, District, Year, Rape, Murder FROM crime_against_st WHERE (Murder) IN ( SELECT MIN(Murder) AS Lowest_murder FROM crime_against_st WHERE District NOT LIKE '%Total%')")
result.fetchall()

[('ANDHRA PRADESH', 'ADILABAD', 2001, 1, 0),
 ('ANDHRA PRADESH', 'ANANTAPUR', 2001, 0, 0),
 ('ANDHRA PRADESH', 'CHITTOOR', 2001, 0, 0),
 ('ANDHRA PRADESH', 'CUDDAPAH', 2001, 0, 0),
 ('ANDHRA PRADESH', 'EAST GODAVARI', 2001, 0, 0),
 ('ANDHRA PRADESH', 'GUNTAKAL RLY.', 2001, 0, 0),
 ('ANDHRA PRADESH', 'GUNTUR', 2001, 2, 0),
 ('ANDHRA PRADESH', 'HYDERABAD CITY', 2001, 0, 0),
 ('ANDHRA PRADESH', 'KARIMNAGAR', 2001, 0, 0),
 ('ANDHRA PRADESH', 'KRISHNA', 2001, 0, 0),
 ('ANDHRA PRADESH', 'KURNOOL', 2001, 0, 0),
 ('ANDHRA PRADESH', 'MEDAK', 2001, 1, 0),
 ('ANDHRA PRADESH', 'NIZAMABAD', 2001, 0, 0),
 ('ANDHRA PRADESH', 'PRAKASHAM', 2001, 0, 0),
 ('ANDHRA PRADESH', 'RANGA REDDY', 2001, 2, 0),
 ('ANDHRA PRADESH', 'SECUNDERABAD RLY.', 2001, 0, 0),
 ('ANDHRA PRADESH', 'SRIKAKULAM', 2001, 0, 0),
 ('ANDHRA PRADESH', 'VIJAYAWADA', 2001, 0, 0),
 ('ANDHRA PRADESH', 'VIJAYAWADA RLY.', 2001, 0, 0),
 ('ANDHRA PRADESH', 'VISAKHA RURAL', 2001, 0, 0),
 ('ANDHRA PRADESH', 'VISAKHAPATNAM', 2001, 0, 0),
 ('ANDHR

Above table shows the records of lowest number of murder case in each state and district and their respective years

## 3.7) Write SQL query to find the number of murders in ascending order in district and yearwise

In [18]:
result=cursor.execute("SELECT state_UT, DISTRICT, Year,Murder FROM crime_against_st GROUP BY DISTRICT ORDER BY Murder ASC, Year ASC")
result.fetchall()

[('WEST BENGAL', '24 PARGANAS NORTH', 2001, 0),
 ('WEST BENGAL', '24 PARGANAS SOUTH', 2001, 0),
 ('ANDHRA PRADESH', 'ADILABAD', 2001, 0),
 ('UTTAR PRADESH', 'AGRA', 2001, 0),
 ('GUJARAT', 'AHMEDABAD COMMR.', 2001, 0),
 ('GUJARAT', 'AHMEDABAD RURAL', 2001, 0),
 ('GUJARAT', 'AHWA-DANG', 2001, 0),
 ('MIZORAM', 'AIZAWL', 2001, 0),
 ('RAJASTHAN', 'AJMER', 2001, 0),
 ('MAHARASHTRA', 'AKOLA', 2001, 0),
 ('KERALA', 'ALAPUZHA', 2001, 0),
 ('UTTAR PRADESH', 'ALLAHABAD', 2001, 0),
 ('UTTARAKHAND', 'ALMORA', 2001, 0),
 ('RAJASTHAN', 'ALWAR', 2001, 0),
 ('HARYANA', 'AMBALA', 2001, 0),
 ('UTTAR PRADESH', 'AMBEDKAR NAGAR', 2001, 0),
 ('MAHARASHTRA', 'AMRAVATI COMMR.', 2001, 0),
 ('MAHARASHTRA', 'AMRAVATI RURAL', 2001, 0),
 ('GUJARAT', 'AMRELI', 2001, 0),
 ('PUNJAB', 'AMRITSAR', 2001, 0),
 ('GUJARAT', 'ANAND', 2001, 0),
 ('ANDHRA PRADESH', 'ANANTAPUR', 2001, 0),
 ('JAMMU & KASHMIR', 'ANANTNAG', 2001, 0),
 ('A & N ISLANDS', 'ANDAMAN', 2001, 0),
 ('ODISHA', 'ANGUL', 2001, 0),
 ('TAMIL NADU', 'ARIYALUR',

#### the table shows the number of murders in ascending order in district and yearwise